obspy/io/scardec/core.py

```diff
@@ -0,0 +1,457 @@
+
+    return ev
+
+
+def _write_scardec(catalog, filename_or_buf, **kwargs):
+    """
+    Write an event to a file.
+
+    :param catalog: The catalog to write. Can only contain one event.
+    :type catalog: :class:`~obspy.core.event.Catalog`
+    :param filename_or_buf: Filename or file-like object to write to.
+    :type filename_or_buf: str, open file, or file-like object.
+    """
+    return _buffer_proxy(filename_or_buf, _internal_write_scardec,
+                         file_mode="wb", catalog=catalog, **kwargs)
+
+
+def _internal_write_scardec(buf, catalog, **kwargs):
+    """
+    Write events to a file.
+
+    :param buf: File to write to.
+    :type buf: Open file or file-like object.
+    :param catalog: The catalog to write.
+    :type catalog: :class:`~obspy.core.event.Catalog`
+    """
+    # Some sanity checks.
+    if len(catalog) < 1:
+        raise ValueError("Catalog must contain at least one event")
+    for event in catalog:
+        _internal_write_single_scardec(buf, event)
+        # Add an empty line between events.
+        if len(catalog) > 1:
+            buf.write(b"\n")
+
+
+def _internal_write_single_scardec(buf, event, **kwargs):
+    """
+    Write an event to a file.
+
+    :param buf: File to write to.
+    :type buf: Open file or file-like object.
+    :param event: The event to write.
+    :type event: :class:`~obspy.core.event.Event`
+    """
+    if not event.focal_mechanisms:
+        raise ValueError("Event must contain a focal mechanism.")
+    foc_mec = event.preferred_focal_mechanism() or event.focal_mechanisms[0]
+    if not event.origins:
+        raise ValueError("Event must have at least one origin.")
+    if not event.magnitudes:
+        raise ValueError("Event must have at least one magnitude.")
+
+    # Attempt to get a moment magnitude
+    mw_candidates = \
+        [_i for _i in event.magnitudes if _i.magnitude_type == "Mw"]
+
+    if not mw_candidates:
+        warnings.warn("No moment wave magnitude found. Will be replaced by the"
+                      " first magnitude in the event object.")
+        mw_mag = event.magnitudes[0]
+    else:
+        mw_mag = mw_candidates[0]
+
+    # Now find the cmt origin. Try to find the first one that is CMT
+    candidates = [_i for _i in event.origins
+                  if _i.origin_type == "centroid"]
+    if candidates:
+        warnings.warn("No derived origin attached to the moment tensor. "
+                      "Will instead use another centroid origin to be "
+                      "written to the file.")
+        cmt_origin = candidates[0]
+    # Otherwise just take the preferred or first one.
+    else:
+        warnings.warn("Could not find a centroid origin. Will instead "
+                      "assume that the preferred or first origin is the "
+                      "centroid origin.")
+        cmt_origin = event.preferred_origin() or event.origins[0]
+
+    if not foc_mec.moment_tensor.source_time_function.extra:
```

Comment:
This won't work I think. If there's no `.extra` this will raise an `AttributeError`. Should be changed to e.g. `if "extra" not in foc_mec.moment_tensor.source_time_function`.